// tribe.storage/sqlite3/initialise.js
var log = require('tribe.logger'),
    Q = require('q'),
    _ = require('underscore');

module.exports = function (entityName, indexes, database) {
    var existingColumns, existingIndexes;

    return loadTableData()
        .then(createTable)
        .then(createIndexes);

    function loadTableData(name) {
        return database.all("pragma table_info(" + entityName + ")")
            .then(function (rows) {
                existingColumns = _.pluck(rows, 'name');
                return database.all("pragma index_list(" + entityName + ")")
            })
            .then(function (rows) {
                existingIndexes = _.pluck(rows, 'name');
            });
    }

    function createTable() {
        if (existingColumns.length === 0)
            return database.run("create table " + entityName + " (__content text)");
    }

    function createIndexes() {
        return Q.all(_.map(indexes, function (index) {
            if (!indexExists(index)) {
                return Q.all(createIndexColumns()).then(createIndex);
            }

            function createIndexColumns() {
                if (index.constructor === Array)
                    return _.map(index, createIndexColumn);
                return [createIndexColumn(index)];
            }

            function createIndexColumn(path) {
                if(!columnExists(path))
                return database.run("alter table " + entityName + " add column " + indexName(path) + " text");
            }

            function createIndex() {
                return database.run("create index " + indexName(index) + " on " + entityName + " (" + indexColumns(index) + ")");
            }

            function indexColumns() {
                if (index.constructor === Array)
                    return _.map(index, indexName).join(',');
                return indexName(index);
            }

            function indexExists() {
                return existingIndexes.indexOf(indexName(index)) > -1;
            }

            function columnExists(path) {
                return existingColumns.indexOf(indexName(path)) > -1;
            }
        }));
    }

    function indexName(index) {
        if (index.constructor === Array)
            return _.map(index, indexName).join('__');
        return index.replace(/\./g, '_');
    }
};
