// tribe.storage/sqlite3/queries.js
var predicateSql = require('./predicateSql'),
    expressions = require('tribe.expressions'),
    _ = require('underscore');

module.exports = {
    store: function (entityData, entity) {
        var columns = _.unique(_.flatten(entityData.indexes));

        if (entityData.keyPath)
            columns.push(entityData.keyPath);

        return _.flatten([{
                sql: "insert or replace into " + entityData.name + " " + columnList() + " values " + paramList(columns.length + 2),
                parameters: paramArray()
            }]
            .concat(columns.map(indexClear))
            .concat(columns.map(indexInsert))
        );

        function indexClear(index) {
            return {
                sql: ''
            }
        }

        function indexInsert(index) {
            return {
                sql: ''
            }
        }

        function columnList() {
            return '(' + ['__content'].concat(columnNames()).join(',') + ')';
        }

        function columnNames() {
            return _.map(columns, columnName);
        }

        function columnName(path) {
            return path.replace(/\./g, '_');
        }

        function paramArray() {
            return [JSON.stringify(entity)].concat(_.map(columns, function (index) {
                return expressions.evaluateKeyPath(index, entity);
            }));
        }
    },
    retrieve: function (entity, predicates) {
        var query = predicateSql.toSql({ name: entity }, predicates);
        query.sql = "SELECT * FROM " + entity + query.sql;
        return query;
    },
    clear: function (entity) {
        return {
            sql: "DELETE FROM " + entity,
            parameters: []
        }
    }
};

function paramList(length) {
    return '(' + new Array(length).join(',?').substring(1) + ')'
}
