// tribe/build/index.js
var activities = require('tribe/build/activities'),
    blocks = require('tribe/build/blocks'),
    templates = require('tribe/build/templates'),
    utils = require('tribe/utilities'),
    Q = require('q'),
    _ = require('underscore'),
    tasks = [],
    options = {};

module.exports = {
    // options: {
    //   tasks: [{ activity: 'activity_name', options: { } }, ...],
    //   phases: ['build', 'phases', ...],
    //   path: 'target/path'
    // }
    configure: function (buildOptions) {
        options = buildOptions;
        tasks = _.map(options.tasks, activities.createTask);
    },

    addTask: function (activity, options) {
        tasks.push(activities.createTask({ activity: activity, options: options }));
    },

    execute: function (context) {
        context = context || {};

        return _.reduce(options.phases, function (lastPhase, phase) {
            return Q.when(lastPhase).then(function () {
                return _.reduce(tasks, function (lastTask, task) {
                    return Q.when(lastTask).then(function () {
                        return task[phase] && task[phase](context);
                    });
                }, null);
            });
        }, null).fail(utils.rethrow('Build failed!'));
    },

    tasks: tasks,
    activities: activities,
    blocks: blocks,
    templates: templates
};
