// tribe/logger.browser.js
var level = 4,
    levels = {
        debug: 4,
        info: 3,
        warn: 2,
        error: 1,
        none: 0
    };

var api = module.exports = {
    setLevel: function (newLevel) {
        level = levels[newLevel];
        if (level === undefined) level = 4;
    },
    debug: function (message) {
        if (level >= 4)
            console.log(('DEBUG: ' + message));
    },
    info: function (message) {
        if (level >= 3)
            console.info(('INFO: ' + message));
    },
    warn: function (message) {
        if (level >= 2)
            console.warn(('WARN: ' + message));
    },
    error: function (message, error) {
        if (level >= 1)
            console.error(('ERROR: ' + message + '\n'), api.errorDetails(error));
    },
    errorDetails: function (ex) {
        if (!ex) return '';
        return (ex.constructor === String) ? ex :
            (ex.stack || '') + (ex.inner ? '\n\n' + this.errorDetails(ex.inner) : '\n');
    },
    log: function (message, prefix) {
        var match = message && message.toString().match(/([^:]*):/),
            level = match && match[1].toLowerCase();

        if (api[level])
            api[level]((prefix ? prefix + ' ' : '') + message.substring(level.length + 2).replace(/\r?\n$/, ''));
        else if (level >= 4)
            console.log((prefix ? prefix + ' ' : '') + message);
    }
};
