// tribe/options.js
var pubsub = require('tribe/pubsub'),
    _ = require('underscore'),
    path = require('path');

pubsub.options.sync = true;
pubsub.options.handleExceptions = false;

module.exports = {
    apply: function (options) {
        _.extend(module.exports, options);
        return module.exports;
    },
    port: 1678,
    appName: 'Tribe',
    basePath: basePath(),
    dataPath: basePath() + 'Data/',
    modulePath: modulePath(),
    testPaths: testPaths(),
    debug: debugMode(),
    debugPort: process.debugPort,
    inspectorPort: 8080,
    childProcess: false,
    build: {
        path: 'Build',
        phases: ['prepare', 'render', 'server'],
        tasks: [
            { activity: 'app' },
            { activity: 'bootstrapper' },
            { activity: 'css' },
            { activity: 'dependencies', options: { path: 'Dependencies' } },
            { activity: 'panes', options: { path: 'Panes' } },
            { activity: 'resources', options: { path: 'Sagas' } },
            { activity: 'resources', options: { path: 'Handlers' } },
            { activity: 'resources', options: { path: 'Services' } },
            { activity: 'scripts', options: { path: 'Scripts' } },
            { activity: 'styles', options: { path: 'Styles' } },
            { activity: 'templates', options: { path: 'Templates' } }
        ]
    },
    test: {
        framework: 'mocha',
        mocha: {
            ui: 'tdd'
        },
        debugPort: 5859,
        restartThrottle: 1000
    }
};

function basePath() {
    var basePath = process.argv[1];
    return basePath.substr(0, basePath.lastIndexOf(path.sep) + 1);
}

function modulePath() {
    return path.join(__dirname, '..');
}

function testPaths() {
    return [
        path.resolve(__dirname, 'tests/server/'),
        basePath() + 'Tests/'
    ];
}

function debugMode() {
    return process.execArgv.indexOf('--debug') > -1 || process.execArgv.indexOf('--debug-brk') > -1;
}
