// tribe/options.js
var pubsub = require('tribe/pubsub'),
    _ = require('underscore'),
    path = require('path');

pubsub.options.sync = true;
pubsub.options.handleExceptions = false;

module.exports = {
    apply: function (options) {
        _.extend(module.exports, options);
        return module.exports;
    },
    port: 1678,
    appName: 'Tribe',
    basePath: defaultBasePath(),
    dataPath: defaultBasePath() + 'Data/',
    modulePath: modulePath(),
    testPaths: testPaths(),
    debug: debugMode(),
    debugPort: process.debugPort,
    inspectorPort: 8080,
    childProcess: false,
    build: {
        phases: ['prepare', 'render', 'server']
    },
    test: {
        framework: 'mocha',
        mocha: {
            ui: 'tdd'
        },
        debugPort: 5859,
        restartThrottle: 1000
    }
};

function defaultBasePath() {
    var basePath = process.argv[1];
    return basePath.substr(0, basePath.lastIndexOf(path.sep) + 1);
}

function modulePath() {
    return path.join(__dirname, '..');
}

function testPaths() {
    return [
        path.resolve(__dirname, 'tests/server/'),
        defaultBasePath() + 'Tests/'
    ];
}

function debugMode() {
    return process.execArgv.indexOf('--debug') > -1 || process.execArgv.indexOf('--debug-brk') > -1;
}
