// tribe/actors/flow.js
// generates an actor definition from an alternate API focused around navigation flows
// the actor instance exposes a single observable called pane. intended to be consumed by the flow binding handler
// this should really be part of composite, but that has not yet been converted to use modules, which adds certain complexities
module.exports = function (definition) {
    return function (actor) {
        var pane = ko.observable()

        this.pane = pane
        actor.to = to

        actor.startsAt = function (startPane, data) {
            pane(createDestination(startPane, data))
        }

        actor.on = function (topic) {
            return {
                to: function (destination, paneData) {
                    actor.handles(topic, function (data) {
                        to(destination)(paneData || data)
                    })
                },
                startChild: function (path, scope, transition) {
                    actor.handles(topic, function (data) {
                        // set the scope to the message data by default - not sure if this is a robust solution
                        actor.startChild(path, scope || data, transition)
                    })
                }
            }
        }

        actor.startChild = function (path, scope, transition) {
            pane({ path: '/__flow', data: { flow: path, transition: transition, scope: scope } })
        }

        definition(actor, to)

        function to(destination, paneData) {
            return function (data) {
                pane(createDestination(destination, paneData || data))
            }
        }
    }

    function createDestination(destination, data) {
        if(destination.constructor === String)
            return { path: destination, data: data }
        return {
            path: destination.path,
            data: destination.data || data,
            scope: destination.scope
        }
    }
}
