// tribe/actors/repository.js
var pubsub = require('tribe.pubsub'),
    actorModule = require('tribe.pubsub/actor'),
    log = require('tribe.logger'),
    _ = require('underscore');

var definitions = {},
    api = module.exports = {
        actors: {},
        actor: function (path) {
            return api.actors[normalisePath(path)];
        },
        register: function (path, constructor) {
            path = normalisePath(path);
            try {
                api.actors[path] = {
                    path: path,
                    constructor: constructor,
                    //definition: extractDefinition(path, constructor)
                };
                log.debug('Registered actor ' + path);
            } catch (ex) {
                log.error('Unable to register actor ' + path, ex);
            }
        },
        definition: function (path) {
            path = normalisePath(path);
            if (!definitions[path])
                definitions[path] = extractDefinition(path, api.actors[path].constructor);
            return definitions[path];
        }
    };

function extractDefinition(path, constructor) {
    var actor = new actorModule(pubsub, constructor);
    return {
        path: path,
        expression: actor.metadata.expression,
        isDistributed: actor.metadata.isDistributed,
        topics: _.keys(actor.handles),
        dependencies: actor.metadata.dependencies
    };
}

function normalisePath(path) {
    return path.charAt(0) !== '/'
        ? '/' + path
        : path;
}
