// tribe/utilities/paths.js
var options = require('tribe/options'),
    path = require('path');

var paths = module.exports = {
    resolveAppPath: function (targetPath) {
        return path.resolve(options.basePath, paths.removeLeadingSlash(targetPath));
    },
    appPathFor: function (fullPath) {
        return fullPath.replace(options.basePath, '').replace(/\\/g, '/');
    },
    resourcePathFor: function (fullPath, resourceFolder) {
        var appPath = module.exports.appPathFor(fullPath);
        return paths.ensureLeadingSlash(appPath.substring(resourceFolder ? resourceFolder.length + 1 : 0, appPath.lastIndexOf('.')));
    },
    markupIdentifierFor: function (path, type) {
        return (type + '-' + path).replace(/\//g, '-').replace(/\./g, '');
    },
    ensureLeadingSlash: function (path) {
        if (path.charAt(0) !== '/')
            return '/' + path;
        return path;
    },
    removeLeadingSlash: function (path) {
        if (path.charAt(0) === '/')
            return path.substring(1);
        return path;
    }
};
