// tribe.graphs/types/incremental.js
var d3 = require('d3')

module.exports = {
    handler: function (value, pathValues, data) {
        // this is a hacky way to show an "incremental" line graph, i.e. all lines update on each message
        if (pathValues.length === 0)
            // unless it's the first message, then just push it on the appropriate array
            pathValues.push(value)
        else
            data.forEach(function(x) {
                if(x === pathValues)
                    x.push(value)
                else
                    x.push(x[x.length - 1] || 0)
            })
    },
    createScaling: function (options) {
        return {
            x: d3.scaleLinear().range([0, options.width]),
            y: d3.scaleLinear().range([options.height, 0])
        }
    },
    createAxes: function (options) {
        return {
            x: d3.axisBottom().tickFormat(d3.format(",.0f")).tickSize(-options.height),
            y: d3.axisLeft().tickFormat(d3.format(",.0f")).tickSize(-options.width)
        }
    },
    createPath: function (scaling) {
        return d3.line().x((d, i) => scaling.x(i)).y(d => scaling.y(d))
    }
}
