// tribe/eventStore/sqlite3/queries.js
var _ = require('underscore');

module.exports = {
    store: function (indexes, envelope) {
        // indexes will eventually support keyPaths instead of just properties on the data object
        return {
            query: "insert into messages " + columnList() + " values " + paramString(),
            params: paramArray()
        };

        function columnList() {
            return '(' + ['id', 'topic', 'envelope'].concat(indexes).join(',') + ')';
        }

        function paramString() {
            return '(' + new Array(indexes.length + 4).join(',?').substring(1) + ')';
        }

        function paramArray() {
            return [envelope.id, envelope.topic, JSON.stringify(envelope)].concat(_.map(indexes, function (index) {
                return envelope.data && envelope.data[index];
            }));
        }
    },
    retrieve: function (options) {
        return {
            query: query(),
            params: params()
        };

        function query() {
            return "select * from messages where " + options.indexName + " = ?" +
                (options.startId ? " and id >= ?" : '');
        }

        function params() {
            var params = [options.indexValue];
            if (options.startId) params.push(options.startId);
            return params;
        }
    }
};
