// tribe/options/index.js
var pubsub = require('tribe/pubsub'),
    _ = require('underscore'),
    path = require('path');

pubsub.options.sync = true;
pubsub.options.handleExceptions = false;

module.exports = {
    apply: function (options) {
        _.extend(module.exports, options);
        return module.exports;
    },
    port: 1678,
    basePath: defaultBasePath(),
    dataPath: defaultBasePath() + 'Data/',
    modulePath: modulePath(),
    testPath: testPath(),
    debug: debugMode(),
    debugPort: process.debugPort,
    inspectorPort: 8080,
    test: {
        framework: 'mocha',
        mocha: {
            ui: 'tdd'
        }
    }
};

function defaultBasePath() {
    var path = process.argv[1];
    return path.substr(0, path.replace(/\\/g, '/').lastIndexOf('/')) + '/';
}

function modulePath() {
    return path.join(__dirname, '..');
}

function testPath() {
    return path.join(__dirname, '../tests/server/');
}

function debugMode() {
    return process.execArgv.indexOf('--debug') > -1 || process.execArgv.indexOf('--debug-brk') > -1;
}
