// tribe.storage/indexeddb/db/store.js
var wrapper = require('./transaction'),
    cursor = require('./cursor');

module.exports = function (db, name) {
    return {
        add: function (entity) {
            return execute(function (store) {
                if (entity.constructor === Array)
                    for (var i = 0, l = entity.length; i < l; i++)
                        store.add(entity[i]);
                else
                    store.add(entity);
            }, true);
        },
        put: function (entity) {
            return execute(function (store) {
                if (entity.constructor === Array)
                    for (var i = 0, l = entity.length; i < l; i++)
                        store.put(entity[i]);
                else
                    store.put(entity);
            }, true);
        },
        single: function (key) {
            var request;
            return execute(function (store) {
                request = store.get(key);
            }).then(function () {
                return request.result;
            });
        },
        clear: function () {
            return execute(function (store) {
                store.clear();
            }, true);
        },
        all: function () {
            return cursor(db, name);
        }
    };

    function execute(callback, writeable) {
        var transaction = db.transaction([name], writeable && 'readwrite'),
            store = transaction.objectStore(name),
            promise = wrapper(transaction);

        callback(store);

        return promise;
    }
};
