// tribe/build/blocks/browserify.js
var utils = require('tribe/utilities').paths,
    options = require('tribe/options'),
    browserify = require('browserify'),
    through = require('through'),
    path = require('path'),
    fs = require('fs'),
    Q = require('q'),
    _ = require('underscore');

module.exports = function (property) {
    var render = {
        to: function (targetPath) {
            return function (context) {
                var b = browserify(),
                    q = Q.defer();

                try {
                    tribeSpecificConfiguration(b, context[property]);

                    _.each(context[property], function (file) {
                        b.add(file.path);
                    });

                    var readStream = b.bundle({ debug: options.debug });

                    if (targetPath) {
                        var writeStream = fs.createWriteStream(utils.resolveAppPath(targetPath));
                        writeStream.on('close', function () {
                            q.resolve();
                        });
                        readStream.pipe(writeStream);
                    } else {
                        // if we don't have a target, return the result through the promise. Makes testing much easier!
                        var content = '';
                        readStream.on('data', function (data) {
                            content += data;
                        });
                        readStream.on('end', function () {
                            q.resolve(content);
                        });
                    }

                    // need to minify with uglify-js if not in debug mode here...
                } catch (ex) {
                    q.reject(ex);
                }
                
                return q.promise;

            };
        },
        toBuild: function (filename) {
            return function (context, options) {
                return render.to(path.join(options.buildPath, filename))(context);
            };
        }
    };
    return render;
};

function tribeSpecificConfiguration(b, files) {
    // these are tribe specific and could be moved to another file with a callback if this needs reusing
    b.add('tribe/client/build/Tribe.debug.js');
    b.add('tribe/client/build/Tribe.Node.Client.debug.js');
    b.require('tribe/register.client', { expose: 'tribe/register' });
    b.transform(setScriptEnvironment);

    function setScriptEnvironment(path) {
        var data = '';
        return through(write, end);

        function write(buf) { data += buf; }
        function end() {
            var file = _.findWhere(files, { path: path });
            if(file) this.queue("T.scriptEnvironment = { resourcePath: '" + file.resourcePath + "' };\n");
            this.queue(data);
            this.queue(null);
        }
    }
}
