// tribe/utilities/index.js
module.exports = {
    ensureTrailingSlash: function (path) {
        return path.charAt(path.length - 1) === '/' ? path : path + '/';
    },
    rethrow: function (message) {
        return function (inner) {
            var error = new Error(message);
            error.inner = inner;
            throw error;
        };
    },
    extractError: function (ex) {
        return ex && ex.constructor === String ?
            { message: ex } :
            {
                message: ex.message,
                name: ex.name,
                stack: ex.stack,
                inner: extractError(ex.inner)
            };
    },
    clearCache: function (path) {
        delete require.cache[fixPath(path)];
        delete require.cache[fixPath(path, true)];
    }
};

function fixPath(path, lower) {
    // sometimes sources of paths return the drive letter in lower case, sometimes upper
    if (path.charAt(1) === ':')
        return (lower ? path.charAt(0).toLowerCase() : path.charAt(0).toUpperCase()) + path.substring(1);
    return path;
}
