// tribe.pubsub/actor.js
var addHandler = require('./actor.addHandler'),
    expressions = require('tribe.expressions'),
    utils = require('./utils');

var actor = module.exports = function (pubsub, definition, scope, dependencies) {
    var self = this;

    pubsub = pubsub.createLifetime(scope);
    this.pubsub = pubsub;
    this.children = [];
    this.scope = scope;
    this.metadata = {
        expression: [],
        scope: []
    };
    this.dependencies = dependencies;
    this.handlers = {};

    configureActor();

    this.topics = function () {
        return utils.keys(this.handlers);
    };

    function configureActor() {
        if (definition && definition.constructor === Function) {
            self.instance = new definition(self);
            self.instance.__actor = self;
        } else {
            //throw new Error("Actor definition must be a function");
            utils.copyProperties(definition, self, ['endsChildrenExplicitly', 'onstart', 'onresume', 'onsuspend', 'onend']);
            self.handlers = definition.handles;
        }
    }
};

actor.prototype.handles = function (topicOrHandlers, handler) {
    if(topicOrHandlers.constructor === String)
        this.handlers[topicOrHandlers] = handler;
    else
        utils.extend(this.handlers, topicOrHandlers);
};

actor.prototype.start = function (startData) {
    if (this.onstart) this.onstart(startData, this);
    utils.each(this.handlers, this.addHandler, this);
    return this;
};

actor.prototype.startChild = function (child, onstartData) {
    this.children.push(new actor(this.pubsub, child)
        .start(onstartData));
    return this;
};

actor.prototype.resume = function (data, resumeData) {
    this.data = data;
    if (this.onresume) this.onresume(resumeData, this);
    utils.each(this.handlers, this.addHandler, this);
    return this;
};

actor.prototype.suspend = function (suspendData) {
    if (this.onsuspend) this.onsuspend(suspendData, this);
    this.pubsub.end();
    this.suspendChildren(suspendData);
    return this;
};

actor.prototype.end = function (endData) {
    if (this.onend) this.onend(endData, this);
    this.pubsub.end();
    this.endChildren(endData);
    return this;
};

actor.prototype.endChildren = function (data) {
    utils.each(this.children, function (child) {
        child.end(data);
    });
};

actor.prototype.suspendChildren = function (data) {
    utils.each(this.children, function (child) {
        child.suspend(data);
    });
};

actor.prototype.isScopedTo = function (property) {
    var self = this;
    utils.each(arguments, function (arg) {
        var value = (self.scope && self.scope.constructor === Object)
            ? self.scope[arg]
            : self.scope;
        self.metadata.expression.push({ p: 'data.' + arg, v: value });
        self.metadata.scope.push(arg);
    });
};

actor.prototype.replay = function (envelopes) {
    var self = this;
    self.pubsub.suspend();
    utils.each(envelopes, function (envelope) {
        if (self.handlers[envelope.topic])
            self.handlers[envelope.topic](envelope.data, envelope);
    });
    self.pubsub.resume();
};

actor.prototype.addHandler = addHandler;
