// tribe.pubsub/lifetime.js
var lifetime = module.exports = function (parent, owner) {
    var self = this,
        channel = require('./channel'),
        actor = require('./actor'),
        utils = require('./utils'),

        tokens = [];

    this.owner = owner;

    this.publish = function(topicOrEnvelope, data) {
        return parent.publish(topicOrEnvelope, data);
    };

    this.publishSync = function(topic, data) {
        return parent.publishSync(topic, data);
    };

    this.subscribe = function (topic, func, filter, options) {
        var token = parent.subscribe(topic, func, filter, options);
        return recordToken(token);
    };

    this.subscribeOnce = function(topic, func) {
        var token = parent.subscribeOnce(topic, func);
        return recordToken(token);
    };
    
    this.unsubscribe = function(token) {
        // we should really remove the token(s) from our token list, but it has trivial impact if we don't
        return parent.unsubscribe(token);
    };

    this.channel = function(channelId) {
        return new channel(self, channelId);
    };

    this.end = function() {
        return parent.unsubscribe(tokens);
    };

    this.createLifetime = function() {
        return new lifetime(self, self.owner);
    };
    
    this.startActor = function (definition, data) {
        return new actor(self, definition).start(data);
    };

    function recordToken(token) {
        if (utils.isArray(token))
            tokens = tokens.concat(token);
        else
            tokens.push(token);
        return token;
    }
};
