// tribe.pubsub/pubsub.js
module.exports = function (instanceOptions) {
    var self = this,
        lifetime = require('./lifetime'),
        channel = require('./channel'),
        actor = require('./actor'),
        topicList = require('./topicList'),
        subscribeOnce = require('./subscribeOnce'),
        subscribeTo = require('./subscribeTo'),
        options = require('./options'),
        utils = require('./utils');

    this.owner = this;
    this.options = instanceOptions || {};
    this.sync = option('sync');
     
    var subscribers = new topicList();
    this.subscribers = subscribers;

    function publish(envelope) {
        var messageSubscribers = subscribers.get(envelope);
        var sync = envelope.sync === true || self.sync === true;

        for (var i = 0, l = messageSubscribers.length; i < l; i++) {
            if (sync)
                executeSubscriber(messageSubscribers[i]);
            else {
                (function (subscriber) {
                    setTimeout(function () {
                        executeSubscriber(subscriber);
                    });
                })(messageSubscribers[i]);
            }
        }

        function executeSubscriber(func) {
            var exceptionHandler = option('exceptionHandler');
            
            if(option('handleExceptions')  && exceptionHandler)
                try {
                    func(envelope.data, envelope);
                } catch (e) {
                    exceptionHandler(e, envelope);
                }
            else
                func(envelope.data, envelope);
        }
    }

    this.publish = function (topicOrEnvelope, data) {
        return publish(createEnvelope(topicOrEnvelope, data));
    };

    this.publishSync = function (topicOrEnvelope, data) {
        var envelope = createEnvelope(topicOrEnvelope, data);
        envelope.sync = true;
        return publish(envelope);
    };
    
    function createEnvelope(topicOrEnvelope, data) {
        return topicOrEnvelope && topicOrEnvelope.topic
            ? topicOrEnvelope
            : { topic: topicOrEnvelope, data: data };
    }

    this.subscribe = function (topic, func, filter, options) {
        if (typeof (topic) === "string")
            return subscribers.add(topic, func, filter, options);
        else if (utils.isArray(topic))
            return utils.map(topic, function(topicName) {
                return subscribers.add(topicName, func, filter, options);
            });
        else
            return utils.map(topic, function (individualFunc, topicName) {
                return subscribers.add(topicName, individualFunc, filter, options);
            });
    };

    this.subscribe.to = subscribeTo(this);

    this.unsubscribe = function (tokens) {
        if (utils.isArray(tokens)) {
            var results = [];
            for (var i = 0, l = tokens.length; i < l; i++)
                results.push(subscribers.remove(tokens[i]));
            return results;
        }

        return subscribers.remove(tokens);
    };

    this.createLifetime = function() {
        return new lifetime(self, self);
    };

    this.channel = function(channelId) {
        return new channel(self, channelId);
    };

    this.startActor = function (definition, data) {
        return new actor(self, definition).start(data);
    };

    this.subscribeOnce = subscribeOnce;
    
    function option(name) {
        return (self.options.hasOwnProperty(name)) ? self.options[name] : options[name];
    }
};
