// tribe/client/validation.js
module.exports = {
    initialize: function () {
        require('knockout.validation')
        require('knockout').validation.init({ 
            parseInputAttributes: true,
            decorateInputElement: true,
            insertMessages: false
        })
    },
    modelValidator: function (model) {
        var isValid = function () {
            model.validators = model.validators || {
                model: ko.validation.group(model),
                fields: ko.validation.group(model.fields, { deep: true })
            }
            model.validators.model.showAllMessages()
            model.validators.fields.showAllMessages()
            return model.validators.model().length === 0 && model.validators.fields().length === 0 
        }
        isValid.reset = function () {
            if(model.validators) {
                model.validators.model.showAllMessages(false)
                model.validators.fields.showAllMessages(false)
            }
        }
        return isValid;
    }
}
