// tribe/server/http.js
var express = require('express'),
    useragent = require('express-useragent'),
    app = express(),
    server = require('http').createServer(app),
    fs = require('q-io/fs'),
    path = require('path'),
    store = require('tribe/store/fs'),
    response = require('./response'),
    options = require('tribe/options'),
    _ = require('underscore');

var api = module.exports = {
    configure: function (modules) {
        app.use(useragent.express());

        _.each(modules, api.addModule);
        return api;
    },
    start: function () {
        // turn this into a service
        app.get('/Data/:partition/:row', function (req, res) {
            store.get(req.params.partition, req.params.row)
                .then(function (data) {
                    res.send(data);
                })
                .fail(function () {
                    res.status(404).end();
                });
        });

        server.listen(options.port);
        return api;
    },
    addModule: function (module) {
        return module.http(app);
    },
    route: function (route) {
        return app.route(route);
    },
    response: response,
    app: app,
    server: server
};
